electroncash/lns.py

```diff
@@ -0,0 +1,555 @@
+            if n_ok:
+                return
+            n_err += 1
+            if n_err < N:
+                return
+        if error_cb:
+            error_cb(exc)
+    def do_lookup_all_staggered():
+        ''' Send req. out to all servers, staggering the requests every 200ms,
+        and stopping early after the first success.  The goal here is to
+        maximize the chance of successful results returned, with tolerance for
+        some servers being unavailable, while also conserving on bandwidth a
+        little bit and not unconditionally going out to ALL servers.'''
+        t0 = time.time()
+        for i, server in enumerate(my_servers):
+            if debug: util.print_error("server:", server, i)
+            lookup_asynch(server,
+                          success_cb = lambda res, _server=server: on_succ(res, _server),
+                          error_cb = lambda exc, _server=server: on_err(exc, _server),
+                          name = name, timeout = timeout,
+                          debug = debug)
+            try:
+                q.get(timeout=0.200)
+                while True:
+                    # Drain queue in case previous iteration's servers also
+                    # wrote to it while we were sleeping, so that next iteration
+                    # the queue is hopefully empty, to increase the chances
+                    # we get to sleep.
+                    q.get_nowait()
+            except queue.Empty:
+                pass
+            with lock:
+                if n_ok:  # check for success
+                    if debug:
+                        util.print_error(f"do_lookup_all_staggered: returning "
+                                         f"early on server {i} of {len(my_servers)} after {(time.time()-t0)*1e3} msec")
+                    return
+    t = threading.Thread(daemon=True, target=do_lookup_all_staggered)
+    t.start()
+
+class LNS(util.PrintError):
+    ''' Class implementing LNS subsystem such as verification, etc. '''
+
+    def __init__(self, wallet):
+        assert wallet, "LNS cannot be instantiated without a wallet"
+        self.wallet = wallet
+        self.lock = threading.Lock()  # note, this lock is subordinate to wallet.lock and should always be taken AFTER wallet.lock and never before
+
+        self._init_data()
+
+        # below is used by method self.verify_name_asynch:
+        self._names_in_flight = defaultdict(list)  # number (eg 100-based-modified height) -> List[tuple(success_cb, error_cb)]; guarded with lock
+
+    def _init_data(self):
+        self.v_by_addr = defaultdict(set) # dict of addr -> set of txid
+        self.v_by_name = defaultdict(set) # dict of lowercased name -> set of txid
+
+    def invalidate(self):
+        ''' This method invalidates the lookup cache and forces the
+            LNS addresses to be refetched after 10 minute interval
+        '''
+        try:
+            QTimer.singleShot(10 * 60 * 1000, self.invalidate) # ten minutes
```

Comment:
Qt is not always available to code living in the `electroncash/` subfolder.  Moreover this is dead code and is never reached. (`start()` is never called).

I will fix this in my commits.